Closes #4556: Passwords sync telemetry support

grigoryk opened the pull request:
This adds a new ping to `support-telemetry-sync` - passwords_sync, which is then used in `services-firefox-accounts` and in `service-sync-logins` to emit password sync telemetry whenever this engine is synchronized. New ping is a verbatim copy of the history_sync ping.

`samples-sync` app was also updated to demonstrate to how synchronize passwords, and how to protect encryption key at rest used for the password storage.

---
<!-- Text above this line will be added to the commit once "bors" merges this PR -->

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

grigoryk commented:
Copied from https://github.com/mozilla-mobile/android-components/pull/3092#issuecomment-499971232, as this telemetry is covering the same system. I've **highlighted** my changes to the original request.

# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

We would like to measure the performance and correctness of our Rust sync implementation. This includes collecting the time taken to sync each data type (currently history and bookmarks **and passwords**), incoming and outgoing record counts, any errors that occur (reporting sanitized error messages in a `string` field), and, for bookmarks, tree structure problem counts.

With the exception of the error string, which does not contain PII, we're submitting timings and counts only.

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?

We need to understand how our new Sync implementation behaves in the wild. The [Sync ping](...) in Firefox Desktop and Firefox for iOS already exists, and has been extremely valuable in identifying and diagnosing Sync issues.

This pull request collects the same information as the Sync ping, but ports its structure to Glean.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

Server-side metrics are not sufficient to understand Sync performance (especially for each step of an engine sync), given that the bulk of the work happens on clients. Validation data can only be collected on the client, since Sync records are encrypted and opaque to the server. The Sync ping for Desktop and iOS provides some stats about Desktop, but, since all three still use different Sync implementations, can't be extrapolated to Fenix.

4) Can current instrumentation answer these questions?

~~Android Components consumers, including Fenix, do not currently report any Sync telemetry.~~
**Android Components consumers, such as Fenix, currently include Sync telemetry for history and bookmarks. There is currently no similar telemetry for passwords.**

5) List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox [data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) on the Mozilla wiki.

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>**Timings, counts, and failure reasons for passwords syncs**</td>
    <td>Interaction data</td>
    <td>https://github.com/mozilla-mobile/android-components/issues/4556</td>
  </tr>
</table>

6) How long will this data be collected?  Choose one of the following:

I want to permanently monitor this data. **(Grisha Kruglov, on behalf of the sync team, e.g. Lina Cambridge et al)**

7) What populations will you measure?

**All users with sync enabled, in products that use service-firefox-accounts and service-sync-logins**
~~All Sync users with Sync enabled.~~

The data is _not_ correlated to the `client_id`; instead, we send a hash of the user's Firefox account ID (`uid`). This does not expose new identifiers, as these are already submitted in the Sync ping on other platforms.

* Which release channels?

**Presumably, all (see individual product owners for details on their sync integration)**.

* Which countries?

**Presumably, all (see individual product owners for details on their sync integration)**.

* Which locales?

**Presumably, all (see individual product owners for details on their sync integration)**.

* Any other filters?  Please describe in detail below.

**Presumably, no (see individual product owners for details on their sync integration)**.

8) If this data collection is default on, what is the opt-out mechanism for users?

Users can opt-out by disabling telemetry, or signing out of Sync.

9) Please provide a general description of how you will analyze this data.

We will create queries in re:dash to monitor the engines, adding to our existing sync engine error/success (https://sql.telemetry.mozilla.org/dashboard/sync-leif-status-dashboard-wip) dashboards and engine error analysis notebook for Desktop (https://gist.github.com/mhammond/66684669e1478d65bd60446cf150c244).

10) Where do you intend to share the results of your analysis?

See above.

11) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?

No.

Dexterp37 commented on components/support/sync-telemetry/metrics.yaml:
```diff
@@ -134,6 +134,132 @@ history_sync:
     expires: never
     lifetime: ping
 
+passwords_sync:
```
In mozilla/glean_parser#146 we'll be restricting names to disallow underscores. Please use dashes (`-`) instead of underscores (`_`).

grigoryk replied:
Will do, thanks! @Dexterp37 should I rename other pings, or leave them as-is?

Dexterp37 replied:
Please leave the other as they are, we added exceptions for them.

grigoryk replied:
:+1:

grigoryk commented:
@Dexterp37 I've addressed your comment regarding naming. This is now blocked on mozilla/glean_parser#146, as otherwise it won't compile since glean-parser currently doesn't consider `passwords-sync` to be a valid name. Do you know when parser support is expected to be updated in a-c?

grigoryk commented:
Glean parser update will land in https://github.com/mozilla-mobile/android-components/pull/5301

grigoryk commented:
Thanks, @liuche. Just waiting for #5301 to land now.

Dexterp37 commented on components/support/sync-telemetry/pings.yaml:
```diff
@@ -26,3 +26,15 @@ bookmarks_sync:
   - sync-core@mozilla.com
   data_reviews:
   - https://github.com/mozilla-mobile/android-components/pull/3092
+#
```
Did you mean to add a comment here?

Dexterp37 commented on components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt:
```diff
@@ -436,19 +436,20 @@ class WorkManagerSyncWorker(
             // so we check for the boolean flag that indicates if this happened or not.
-            // There's a complete mismatch between what Glean supports and what we need
-            // it to do here. They don't support "nested metrics" and so we resort to these hacks.
+            // There's a complete mismatch between what Glean supports and what we need it to do here.
```
This is a bit unfair, to be honest.

The sync team, the Glean teams and DS [designed and agreed](https://docs.google.com/document/d/1aJnqqsLnhIAzEJR4UylXO8jUVrkw68ZMGHMjKTANtQA/edit#heading=h.9qtbe4mxitxr) on the solution to send "one ping per engine sync". This is not, IMHO, a shortcoming of the SDK: this is doing what we all agreed on.

We can revisit the decision, of course.

grigoryk commented on components/support/sync-telemetry/metrics.yaml:
```diff
@@ -134,6 +134,135 @@ history_sync:
+# Note that this ping is named using kebab-case.
+# snake_case is no longer supported by Glean, see https://bugzilla.mozilla.org/show_bug.cgi?id=1601270.
+# history_sync and bookmarks_sync pings are kept as-is, as there are specific exceptions for them in the glean parser.
+passwords-sync:
```
"Logins" is a more encompassing name, so let's go with that. At this point we have a pretty even split between "logins" and "passwords" in our codebase, all largely referring to the same thing :)

grigoryk commented:
Thanks for the reviews all. @Dexterp37 my apologies for that comment - it was written from a somewhat myopic point of view, which didn't take into account the larger picture.

grigoryk commented:
bors r=rocketsroger,liuche

Dexterp37 commented:
> 
> @Dexterp37 my apologies for that comment - it was written from a somewhat myopic point of view, which didn't take into account the larger picture.

No problem! I'm glad we're all aligned now :-)
